plugin.audio.soundcloud/resources/lib/soundcloud/api_v2.py

```diff
@@ -0,0 +1,238 @@
+from future import standard_library
+standard_library.install_aliases()  # noqa: E402
+
+import hashlib
+import json
+import requests
+import urllib.parse
+import xbmc
+
+from resources.lib.models.playlist import Playlist
+from resources.lib.models.track import Track
+from resources.lib.models.selection import Selection
+from resources.lib.models.user import User
+from resources.lib.soundcloud.api_collection import ApiCollection
+from resources.lib.soundcloud.api_interface import ApiInterface
+
+
+class ApiV2(ApiInterface):
+    """This class uses the unofficial API used by the SoundCloud website."""
+
+    api_host = "https://api-v2.soundcloud.com"
+    api_client_id = "KT6UCHXC9iNnI8wn4UUfwMSlAPe4Z8zx"
```

Comment:
This API key changes regularly, I don't know how long it stays valid. But the key can be overridden via settings.

The key should be automatically extracted from the SoundCloud website. Otherwise the plugin has to be updated as soon as the key expires.

Reply:
Is there a hook, so I can extract the key on every Kodi start-up?

Reply:
I think this can be done with a service (https://kodi.wiki/view/Service_add-ons). At least Twitch is doing it like this.

Reply:
I implemented it in https://github.com/jaylinski/kodi-addon-soundcloud/pull/9 and will push it upstream as soon as I receive the first feedback.

Reply:
Only did some quick googeling and to me it seems as v2 would require you to implement oauth to officially get authentication related keys (which would include refresh-tokens etc.)

This would be a little bit more involved but also be a cleaner solution.

If I'm missing something please let me know as I really didn't do much research on this.

P.s. I raised this internally to see if other are okay with this solution (for now/indefinitely)

Reply:
You can create a service extension point to extract the new key on kodi on specific time-intervals

Reply:
Another option is to decorate the methods that require the key to a request (prior to the method execution) which will grab the key and cache it. If the last request don't exceed a specific threshold just use the cached one. If the request fails with auth-related http status codes just grab it again, and so on

Reply:
Even better, make `api_client_id` a property which handles the caching.
Then other functions don't need decorators and it's a easy drop-in replacement.

Something like:
```python
...
@property
def api_client_id(self):
  if <should_fetch_id_again_because_enough_time_passed>:
    # fetch key
    self._api_client_id = <new id>
    # write to cache

  if not self._api_client_id:
    # self._api_client_id = <read from cache>

  return self._api_client_id
...
```

Reply:
@enen92 @tadly thanks for your feedback!
I will try your proposed approaches in the future (https://github.com/jaylinski/kodi-addon-soundcloud/issues/11). If it's ok for you, I'll stick with the service-approach for now.

Regarding the cleaner approach with the official oauth implementation: SoundCloud locked the app-registration about a year ago: http://soundcloud.com/you/apps/new
So it's impossible to use the official API if you didn't register an app a year ago... :man_shrugging:

Reply:
@jaylinski ah. Well that's a bummer but good to know :)